### ISSUE TITLE
sys.setdefaultencoding() usage in twisted.test.test_log makes portability hard

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3000]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3000|
|Type|defect|
|Created|2008-01-19 17:46:32Z|
|Branch|https://github.com/twisted/twisted/tree/conditional-setdefaultencoding-3000|

```sys.setdefaultencoding``` is a bad API.  CPython basically gave up on it when it changed ```site.py``` to delete the attribute.  The only way to actually use it is with a ```reload``` hack that brings the attribute back.  This is somewhat deep hackery.

```twisted.test.test_log``` uses this API to change the default encoding to _ascii_ so that it can test the behavior of ```twisted.python.log``` when handling certain encoding errors.  However, this usage makes the tests unportable to alternate Python runtimes which don't respond in the same way to the hackery used to expose the API (eg, on `PyPy`, ```reload``` does not bring back the attribute).

The usage can be made conditional on the current default encoding.  This way, runtimes which do not allow the default encoding to be changed from ascii will not need to provide the API, but the tests will still exercise the same functionality.

<details><summary>Searchable metadata</summary>

```
trac-id__3000 3000
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_conditional_setdefaultencoding_3000 branches/conditional-setdefaultencoding-3000
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__pypy pypy
time__1200764792000000 1200764792000000
changetime__1200836715000000 1200836715000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [22352]) Branching to 'conditional-setdefaultencoding-3000'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

The test now checks to see if str(non-ascii) raises an exception and only calls setdefaultencoding if it does not.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [22371]) Merge conditional-setdefaultencoding-3000

Author: exarkun
Reviewer: therve
Fixes [#3000](https://github.com/twisted/twisted/issues/3000)

Change the unit tests for twisted.python.log so that they only call
sys.setdefaultencoding if it is necessary in order to test the behavior
of logging unicode strings.  That is, only change the default encoding
if it is not already ASCII.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
